### DISCUSSION TITLE
StreamingResponse of opencv2 frames in bytes for multiple clients

### DISCUSSION THREAD

**Discussion by JustinYeoh93** [fastapi/fastapi#6726] [Questions] [answered]
### First check

* [X] I added a very descriptive title to this issue.
* [X] I used the GitHub search to find a similar issue and didn't find it.
* [X] I searched the FastAPI documentation, with the integrated search.
* [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
* [X] I already read and followed all the tutorial in the docs and didn't find an answer.
* [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
* [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
* [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).
* [X] After submitting this, I commit to one of:
    * Read open issues with questions until I find 2 issues where I can help someone and add a comment to help there.
    * I already hit the "watch" button in this repository to receive notifications and I commit to help at least 2 people that ask questions in the future.
    * Implement a Pull Request for a confirmed bug.

### Example

Here's a self-contained, [minimal, reproducible, example](https://stackoverflow.com/help/minimal-reproducible-example) with my use case:

```Python
from fastapi import FastAPI, Depends
from fastapi.responses import StreamingResponse
import cv2

app = FastAPI()

def camera_setup():
    vid = cv2.VideoCapture(0)
    return vid

@app.get('/')
async def video_feed(vid=Depends(camera_setup)):
    return StreamingResponse(generate(vid), media_type="multipart/x-mixed-replace;boundary=frame")

def generate(vid):
    try:
        while True:
            fet, frame = vid.read()
            binary_string = cv2.imencode('.png', frame)[1].tobytes()
            yield (b'--frame\r\n' b'Content-Type: image/jpeg\r\n\r\n' +
                        binary_string + b'\r\n') 
    except:
        print("Oh no, someone disconnected")

if __name__ == "__main__":
    import uvicorn
    uvicorn.run("camera:app", reload=True)
```

### Description

* Open the browser and call the endpoint `/`.
* It returns a video of the camera 
* But when I open another instance, the first instance gets disconnected.
* What I hoped was that the first and second frontends will continue streaming the contents from the backend.

### Environment

* OS: Windows10
* FastAPI Version [e.g. 0.3.0]: 0.63.0
* Python version: 3.7.9

### Additional context
* I have tried it on Quartz and Flask and they do not suddenly just stop
* Appreciate y'all taking the time to help

**Comment by JustinYeoh93** [comment, marked as answer]
Ok, I just did a little bit of looking at the error messages and found out that this may be a opencv problem
On opening a new tab with the address, the below error surfaces.

```bash
[ WARN:1] global C:\Users\appveyor\AppData\Local\Temp\1\pip-req-build-oduouqig\opencv\modules\videoio\src\cap_msmf.cpp (434) `anonymous-namespace'::SourceReaderCB::~SourceReaderCB terminating async callback
```

I am assuming that because I have initialized the camera again, and somehow that caused the camera to disappear.

**Comment by JustinYeoh93** [comment]
I feel this question might be a little skewed from FastAPI and more to do with the architecture of my application.
Appreciate those who had took their time to look at this. 
I shall close this due to irrelevancy.

PS: keep up the great work!

**Comment by tiangolo** [comment]
Thanks for reporting back and closing the issue 👍

Yep, I'm pretty sure OpenCV will allow you to capture the frames only once at one point. If you need to share and stream those frames to multiple requests, you probably need a more complex setup. You could save the frames to a Redis queue, or if you need to do it locally in memory on the same code, you could put them on an AnyIO stream, and then read them from multiple places subscribed to the same stream, but then that code would be quite more complex.

And you would have to write the code to combine blocking code (OpenCV) with async code (AnyIO), you might want to check Asyncer to help you with that: https://asyncer.tiangolo.com/

> Sorry for the long delay! 🙈  I wanted to personally address each issue/PR and they piled up through time, but now I'm checking each one in order.
